contracts/liquidityMining/KyberFairLaunch.sol

```diff
@@ -0,0 +1,469 @@
+  // rewardPerBlock: amount of reward token per block
+  // accRewardPerShare: accumulated reward per share of token
+  struct PoolInfo {
+    address stakeToken;
+    uint32  startBlock;
+    uint32  endBlock;
+    uint32  rewardLockBps;
+    uint32  lastRewardBlock;
+    uint128 rewardPerBlock;
+    uint128 accRewardPerShare;
+  }
+
+  // contract for locking reward
+  IKyberRewardLocker public immutable rewardLocker;
+  // reward token
+  IERC20Ext public immutable rewardToken;
+
+  // Info of each pool.
+  PoolInfo[] public poolInfo;
+  // Info of each user that stakes Staking tokens.
+  mapping(uint256 => mapping(address => UserInfo)) public userInfo;
+
+  event AddNewPool(
+    address indexed stakeToken,
+    uint32 indexed startBlock,
+    uint32 indexed endBlock,
+    uint32 rewardLockBps,
+    uint256 rewardPerBlock
+  );
+  event RenewPool(
+    uint256 indexed pid,
+    uint32 indexed startBlock,
+    uint32 indexed endBlock,
+    uint32 rewardLockBps,
+    uint256 rewardPerBlock
+  );
+  event UpdatePool(
+    uint256 indexed pid,
+    uint32 indexed endBlock,
+    uint32 rewardLockBps,
+    uint256 rewardPerBlock
+  );
+  event Deposit(
+    address indexed user,
+    uint256 indexed pid,
+    uint256 indexed blockNumber,
+    uint256 amount
+  );
+  event Withdraw(
+    address indexed user,
+    uint256 indexed pid,
+    uint256 indexed blockNumber,
+    uint256 amount
+  );
+  event Harvest(
+    address indexed user,
+    uint256 indexed pid,
+    uint256 indexed blockNumber,
+    uint256 amount,
+    uint256 lockedAmount
+  );
+  event EmergencyWithdraw(
+    address indexed user,
+    uint256 indexed pid,
+    uint256 indexed blockNumber,
+    uint256 amount
+  );
+
+  constructor(
+    address _admin,
+    IERC20Ext _rewardToken,
+    IKyberRewardLocker _rewardLocker
+  ) PermissionAdmin(_admin) {
+    rewardToken = _rewardToken;
+    rewardLocker = _rewardLocker;
+
+    _rewardToken.safeApprove(address(_rewardLocker), uint256(-1));
+  }
+
+  /**
+  *  @dev allow admin to withdraw only reward token
+  */
+  function adminWithdraw(uint256 amount) external onlyAdmin {
+    rewardToken.safeTransfer(msg.sender, amount);
+  }
+
+  /**
+  * @dev Add a new lp to the pool. Can only be called by the admin.
+  * @param _stakeToken: token to be staked to the pool
+  * @param _startBlock: block where the reward starts
+  * @param _endBlock: block where the reward ends
+  * @param _rewardLockBps: percentage (in bps) of reward to be locked
+  * @param _rewardPerBlock: amount of reward token per block for the pool
+  * @param _withUpdate: whether to update all pools
+  */
+  function addPool(
+    address _stakeToken,
+    uint32 _startBlock,
+    uint32 _endBlock,
+    uint32 _rewardLockBps,
+    uint128 _rewardPerBlock,
+    bool _withUpdate
+  ) external override onlyAdmin {
+    if (_withUpdate) {
+      massUpdatePools();
+    }
+    require(_stakeToken != address(0), 'add: not stakeToken addr');
```

Comment:
hmmm i'm shifting towards removing null address checks, because
1) from experience, we've not made that mistake
2) `0x0` is simply one of many invalid / wrong addresses
3) save gas, might be cheap now, but if eth hits 5k and we see network congestion again... no longer cheap :p

Reply:
Remove checked, seems redundant anw lol